### DISCUSSION TITLE
getInitialProps Circular structure error, but I'm not using getInitialProps!

### DISCUSSION THREAD

**Discussion by jrr** [vercel/next.js#47135] [Help]
### Summary

I have a page that uses `getServerSideProps`, _not_ `getInitialProps`.

I'm getting the following error:

```
Error: Circular structure in "getInitialProps" result of page "/resources/[slug]". https://nextjs.org/docs/messages/circular-structure
```

I deliberately avoided returning circular data from `getServerSideProps`, because it is known not to work with the underlying JSON serialization. I know that the server props are okay, because putting a cycle in there causes a different error.

I _do_ still have circular data produced during rendering, and passed around as props, but from what I've gleaned I thought that was okay.

Is this actually _not_ okay? Are component props also subject to the JSON-serializable restriction?

Could the error message be misleading? (Are there other circumstances that could provoke it, apart from using getInitialProps? Does this check somehow get applied to server-side rendering of getServerSideProps pages?)

Any tips for debugging where this `getInitialProps` invocation is happening, if not from application code?

Here's a screenshot of the error with the top few frames of the call stack:

![image](https://user-images.githubusercontent.com/164652/225177744-720e0d98-dbcc-4497-ac9e-5fc585b8867b.png)

I've not yet been able to create an isolated repro, but if I can I'll certainly share it.

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:37 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.15.0
      npm: 9.5.0
      Yarn: N/A
      pnpm: 7.29.1
    Relevant packages:
      next: 13.2.5-canary.3
      eslint-config-next: 13.2.4
      react: 18.2.0
      react-dom: 18.2.0
```
```


### Example

_No response_

**Comment by icyJoseph** [comment]
This seems to be failing when Next.js generates the `__NEXT_DATA__` script tag, which has `type="application/json"`, this is used client side to rehydrate the application.

The contents of that look like this:

```json
{
  "props": {
    "pageProps": {},
    "__N_SSG": true
  },
  "page": "/",
  "query": {},
  "buildId": "jhEZCWSrWH_BpgInL76sr",
  "isFallback": false,
  "dynamicIds": [304],
  "gsp": true,
  "scriptLoader": []
}
```

We would expect the error to be in pageProps though. Perhaps through mutation you are accidentally creating a circular dependency?

**Reply by icyJoseph** [reply]
Also, I have seen someone with a similar issue before, but couldn't really pin the cause: https://github.com/vercel/next.js/discussions/46215#discussioncomment-5103833

**Reply by jrr** [reply]
@icyJoseph thanks for the tip! With a better idea of what to look for, I found some mutation happening to the page props. Here's a minimal repro:

```tsx

/*
  If you mutate page props during rendering and add a cycle, Next will throw a misleading error:

  Error: Circular structure in "getInitialProps" result of page "/example". https://nextjs.org/docs/messages/circular-structure
*/
import { InferGetServerSidePropsType } from "next";

export const getServerSideProps = async ({ req, res }) => {
  return { props: { message: "hello", foo: {} } };
};

type Props = InferGetServerSidePropsType<typeof getServerSideProps>;

const ExamplePage: React.FC<Props> = (props) => {
  (props.foo as any).baz = props.foo;
  return <>{props.message}</>;
};

export default ExamplePage;
```

I'll plan to file an issue on this. I'm not sure if it would be possible for Next to separately detect problems of this form, but at minimum the error message could be reworded to cover this other situation that may provoke the error.

**Reply by icyJoseph** [reply]
The thing is that all pages have an implicit `getInitialProps`, you just don't see it, unless you implement your own... sort of... I am making huge simplifications here - At the end of the day, this is used to fill up the `__NEXT_DATA__` tag, for sure, it could be better done, but I am not sure there's enough information at that point, all they do is `JSON.stringify(something)` and that can fail in two ways: 

- Circular Data Structure
- BigInt

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#exceptions

I'll look into the snippet you shared.

**Reply by icyJoseph** [reply]
@jrr Interesting snippet, I've noticed one thing.

If you are at page A, which doesn't do any of this, and navigate, using SPA transitions, to page B, which does mutate the props, everything works fine.

That's because the response from `GSSP`, is serialised and transferred over the network to the app, and that in turn is used client side as props.

However, if you navigate directly to page B, or refresh the page there, it crashes, because the entire page is made server side, and presumably, the creation of `__NEXT_DATA__` happens after the component has been rendered, but both are using the same reference, which is the result of `GSSP`. Fun stuff.

If you were to use:

```js
export const getServerSideProps = async () => {
  return { props: { message: "hello", foo: Object.freeze({}) } };
};
```

Then, the direct navigation step, fails at `TypeError: Cannot add property baz, object is not extensible`, which would mean that indeed, this step did happen first (since we don't see the circular error just yet).

**Reply by jrr** [reply]
> The thing is that all pages have an implicit getInitialProps, you just don't see it

☝️ This would have been immensely helpful when I started digging into this! It'd be nice to have this described in docs and error messages.

**Reply by icyJoseph** [reply]
Yeah, well, it is a simplification, but still the origin of the issue. Had to navigate through the code. If we get technical I think that statement may fall apart, 😅
